File: src/api/agent.ts
import { Channel, invoke } from "@tauri-apps/api/core";
import type { ModArchiveImportOutcome } from "./modLibrary";

export type DeepSeekModel = "v4Flash" | "v4Pro";

export interface AgentSettings {
  model: DeepSeekModel;
  modelApiName: string;
  apiKeyConfigured: boolean;
  apiKeyHint: string | null;
  apiKeySource: "credentialManager" | "environment" | null;
}

export interface AgentConnectionResult {
  model: DeepSeekModel;
  modelApiName: string;
  elapsedMillis: number;
  message: string;
}

export interface AgentActionTarget {
  modId: string;
  name: string;
  detail: string;
}

export interface AgentActionPlan {
  planId: string;
  kind:
    | "batchModAction"
    | "conflictOrder"
    | "modelRemap"
    | "cleanupExclusions"
    | "cleanupRestore";
  title: string;
  summary: string;
  stateVersion: string;
  expiresAtUnixSeconds: number;
  targetCount: number;
  targets: AgentActionTarget[];
  warnings: string[];
  destructive: boolean;
}

export type CleanupRecommendation = "remove" | "review" | "keep";

export interface AgentCleanupReviewItem {
  candidateId: string;
  modId: string;
  modName: string;
  libraryRelativePath: string;
  deployRelativePath: string;
  sizeBytes: number;
  currentlyDeployed: boolean;
  recommendation: CleanupRecommendation;
  reason: string;
  confidence: number;
  selectedByDefault: boolean;
  decisionSource: "localRule" | "acuAi";
  riskLevel: "low" | "medium" | "high";
}

export interface AgentCleanupReview {
  reviewId: string;
  candidateCount: number;
  scannedFileCount: number;
  localKeepCount: number;
  localRemoveCount: number;
  aiReviewCount: number;
  aiGroupCount: number;
  ruleVersion: number;
  items: AgentCleanupReviewItem[];
  message: string;
}

export interface AgentActionResult {
  planId: string;
  status: "completed" | "partiallyFailed" | "cancelled";
  title: string;
  message: string;
  succeededCount: number;
  failedCount: number;
  warnings: string[];
  archiveImport: ModArchiveImportOutcome | null;
}

export type AgentEventKind =
  | "started"
  | "textReset"
  | "textDelta"
  | "toolStarted"
  | "toolFinished"
  | "planReady"
  | "cleanupReviewReady"
  | "knowledgeEvidenceReady"
  | "completed"
  | "failed";

export interface AgentKnowledgeEvidence {
  evidenceId: string;
  title: string;
  gameVersion: string;
  confidence: number;
  sourceTitle: string | null;
  sourceUrl: string | null;
  packId: string;
  packVersion: string;
  sourceTier: "localVerified" | "localReference" | "localAnalysis" | "webReference";
}

export interface AgentWebSearchTestResult {
  model: DeepSeekModel;
  modelApiName: string;
  elapsedMillis: number;
  searchResultCount: number;
  pageReadSucceeded: boolean;
  source: string | null;
  message: string;
}

export interface AgentEvent {
  turnId: string;
  sequence: number;
  kind: AgentEventKind;
  text: string | null;
  toolName: string | null;
  message: string | null;
  plan: AgentActionPlan | null;
  cleanupReview: AgentCleanupReview | null;
  knowledgeEvidence: AgentKnowledgeEvidence[];
}

export interface AgentTurnResult {
  turnId: string;
  message: string;
}

export function getAgentSettings(): Promise<AgentSettings> {
  return invoke<AgentSettings>("get_agent_settings");
}

export function saveAgentModel(model: DeepSeekModel): Promise<AgentSettings> {
  return invoke<AgentSettings>("save_agent_model", { model });
}

export function setDeepSeekApiKey(apiKey: string): Promise<AgentSettings> {
  return invoke<AgentSettings>("set_deepseek_api_key", { apiKey });
}

export function deleteDeepSeekApiKey(): Promise<AgentSettings> {
  return invoke<AgentSettings>("delete_deepseek_api_key");
}

export function testAgentConnection(): Promise<AgentConnectionResult> {
  return invoke<AgentConnectionResult>("test_agent_connection");
}

/** 使用已保存的 Key 验证 DeepSeek 服务端搜索和白名单页面读取。 */
export function testAgentWebSearch(): Promise<AgentWebSearchTestResult> {
  return invoke<AgentWebSearchTestResult>("test_agent_web_search");
}

/** Channel 保证同一轮流式片段有序到达，前端不接触 DeepSeek 原始 SSE。 */
export function startAgentTurn(
  message: string,
  handleEvent: (event: AgentEvent) => void,
): Promise<AgentTurnResult> {
  const onEvent = new Channel<AgentEvent>();
  onEvent.onmessage = handleEvent;
  return invoke<AgentTurnResult>("start_agent_turn", { message, onEvent });
}

export function clearAgentSession(): Promise<void> {
  return invoke<void>("clear_agent_session");
}

export function confirmAgentActionPlan(planId: string): Promise<AgentActionResult> {
  return invoke<AgentActionResult>("confirm_agent_action_plan", { planId });
}

export function cancelAgentActionPlan(planId: string): Promise<AgentActionResult> {
  return invoke<AgentActionResult>("cancel_agent_action_plan", { planId });
}

export function createAgentCleanupPlan(
  reviewId: string,
  candidateIds: string[],
): Promise<AgentActionPlan> {
  return invoke<AgentActionPlan>("create_agent_cleanup_plan", { reviewId, candidateIds });
}

File: src-tauri/src/commands/agent.rs
use tauri::{ipc::Channel, State};

use crate::{
    services::agent::{
        self, AgentActionPlan, AgentActionResult, AgentConnectionResult, AgentCoordinator,
        AgentEvent, AgentSettings, AgentTurnResult,
    },
    services::nexus::NexusConnectionResult,
    storage::config::DeepSeekModel,
};

/// 读取 DeepSeek V4 模型和脱敏凭据状态。
#[tauri::command]
pub fn get_agent_settings(app: tauri::AppHandle) -> Result<AgentSettings, String> {
    agent::get_agent_settings(&app)
}

/// 保存非敏感的 DeepSeek V4 模型选择。
#[tauri::command]
pub fn save_agent_model(
    app: tauri::AppHandle,
    model: DeepSeekModel,
) -> Result<AgentSettings, String> {
    agent::save_agent_model(&app, model)
}

/// 将 DeepSeek 访问密钥保存到系统凭据库，普通配置不会保存明文。
#[tauri::command]
pub fn set_deepseek_api_key(
    app: tauri::AppHandle,
    api_key: String,
) -> Result<AgentSettings, String> {
    agent::set_deepseek_api_key(&app, api_key)
}

/// 删除 Acumod 写入系统凭据库的 DeepSeek 访问密钥。
#[tauri::command]
pub fn delete_deepseek_api_key(app: tauri::AppHandle) -> Result<AgentSettings, String> {
    agent::delete_deepseek_api_key(&app)
}

/// 使用当前 Key 和模型发送最小请求，验证 DeepSeek 服务是否可用。
#[tauri::command]
pub async fn test_agent_connection(app: tauri::AppHandle) -> Result<AgentConnectionResult, String> {
    agent::test_agent_connection(&app).await
}

/// 将 Nexus Personal API Key 保存到系统凭据库。
#[tauri::command]
pub fn set_nexus_api_key(app: tauri::AppHandle, api_key: String) -> Result<AgentSettings, String> {
    agent::set_nexus_api_key(&app, api_key)
}

/// 删除 Acumod 写入系统凭据库的 Nexus Personal API Key。
#[tauri::command]
pub fn delete_nexus_api_key(app: tauri::AppHandle) -> Result<AgentSettings, String> {
    agent::delete_nexus_api_key(&app)
}

/// 验证 Nexus API Key、账户名称和 API 直接下载权限。
#[tauri::command]
pub async fn test_nexus_connection() -> Result<NexusConnectionResult, String> {
    agent::test_nexus_connection().await
}

/// 开始一次 Agent 对话；写操作只能生成等待确认的受控计划。
#[tauri::command]
pub async fn start_agent_turn(
    app: tauri::AppHandle,
    coordinator: State<'_, AgentCoordinator>,
    message: String,
    on_event: Channel<AgentEvent>,
) -> Result<AgentTurnResult, String> {
    agent::start_agent_turn(app, coordinator.inner().clone(), message, on_event).await
}

/// 重新校验当前状态并执行一份尚未过期的 AI 操作计划。
#[tauri::command]
pub async fn confirm_agent_action_plan(
    app: tauri::AppHandle,
    coordinator: State<'_, AgentCoordinator>,
    plan_id: String,
) -> Result<AgentActionResult, String> {
    agent::confirm_agent_action_plan(app, coordinator.inner().clone(), plan_id).await
}

/// 取消并销毁一份 AI 操作计划，不修改 MOD 状态。
#[tauri::command]
pub fn cancel_agent_action_plan(
    coordinator: State<'_, AgentCoordinator>,
    plan_id: String,
) -> Result<AgentActionResult, String> {
    agent::cancel_agent_action_plan(coordinator.inner(), plan_id)
}

/// 根据用户在清理建议中的逐项选择生成待确认计划，不执行文件操作。
#[tauri::command]
pub async fn create_agent_cleanup_plan(
    app: tauri::AppHandle,
    coordinator: State<'_, AgentCoordinator>,
    review_id: String,
    candidate_ids: Vec<String>,
) -> Result<AgentActionPlan, String> {
    let worker_app = app.clone();
    let worker_coordinator = coordinator.inner().clone();
    tauri::async_runtime::spawn_blocking(move || {
        agent::create_agent_cleanup_plan(&worker_app, &worker_coordinator, review_id, candidate_ids)
    })
    .await
    .map_err(|error| format!("清理计划生成任务失败：{error}"))?
}

/// 清空当前运行期间的 AI 会话，不修改任何 MOD 状态。
#[tauri::command]
pub fn clear_agent_session(coordinator: State<'_, AgentCoordinator>) -> Result<(), String> {
    coordinator.clear()
}
